### PR TITLE
23 add expense period

### LINKED ISSUE #23: Add expense period
An expense period is an expense that the app add automatically after x time, where x is a period chose by a client. 

- [x] Add expense period 
- [x] Make expense period working

### PR DESCRIPTION
**Pull request by simo2898**

### CODE REVIEW

File: src/main/java/com/expensetracker/cronjobs/ExpensesCronJobs.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import com.expensetracker.service.ExpensePeriodService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.PostConstruct;
+import java.util.List;
+
+@RequiredArgsConstructor
+@Component
+public class ExpensesCronJobs {
```
Comment (mattiapitossi) [inline src/main/java/com/expensetracker/cronjobs/ExpensesCronJobs.java]: Should we change this to a service?

File: src/main/java/com/expensetracker/mapper/impl/ExpenseMapperImpl.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package com.expensetracker.mapper.impl;
+
+import com.expensetracker.mapper.ExpenseMapper;
+import com.expensetracker.model.Expense;
+import com.expensetracker.model.ExpensePeriod;
+import org.springframework.stereotype.Component;
+
+@Component
```
Comment (mattiapitossi) [inline src/main/java/com/expensetracker/mapper/impl/ExpenseMapperImpl.java:8]: There should be an annotation called @Mapper

File: src/main/java/com/expensetracker/service/impl/ExpensePeriodServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            setExpenseNextPayment(expensePeriod);
+            expensePeriodRepository.save(expensePeriod);
+        }
+    }
+
+    private void setExpenseNextPayment(ExpensePeriod expensePeriod) {
+        LocalDate nextPayment = null;
+
+        switch (expensePeriod.getPeriodType()) {
+            case "YEAR":
```
Comment (mattiapitossi) [inline src/main/java/com/expensetracker/service/impl/ExpensePeriodServiceImpl.java]: You could also use the enhanced Switch/Case from Java 17 so you don't need to use the break.

File: src/main/java/com/expensetracker/service/impl/ExpensePeriodServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                nextPayment = expensePeriod.getNextPayment().plusYears(expensePeriod.getPeriodDate());
+                break;
+            case "MONTH":
+                nextPayment = expensePeriod.getNextPayment().plusMonths(expensePeriod.getPeriodDate());
+                break;
+            case "DAY":
+                nextPayment = expensePeriod.getNextPayment().plusDays(expensePeriod.getPeriodDate());
+                break;
+        }
```
Comment (mattiapitossi) [inline src/main/java/com/expensetracker/service/impl/ExpensePeriodServiceImpl.java]: Default?

**Review by mattiapitossi — CHANGES_REQUESTED** [review]

File: src/main/java/com/expensetracker/mapper/impl/ExpenseMapperImpl.java (hunk 2 above)
Comment (simo2898) [reply src/main/java/com/expensetracker/mapper/impl/ExpenseMapperImpl.java:8]: The annotation comes from a dependency called org.mapstruct

**Review by mattiapitossi — APPROVED** [review]

**Pull request merged**